README.md
# jq-rs

[![crates.io](https://img.shields.io/crates/v/jq-rs.svg)](https://crates.io/crates/jq-rs)
[![crates.io](https://img.shields.io/crates/d/jq-rs.svg)](https://crates.io/crates/jq-rs)
[![docs.rs](https://docs.rs/jq-rs/badge.svg)](https://docs.rs/jq-rs)
[![Build Status](https://github.com/onelson/jq-rs/actions/workflows/test.yaml/badge.svg)](https://github.com/onelson/jq-rs/actions/workflows/test.yaml)

## Overview

> Prior to v0.4.0 this crate was named [json-query].

This rust crate provides access to [jq] 1.6 via the `libjq` C API (rather than
"shelling out").

By leveraging [jq] we can extract data from json strings using `jq`'s dsl.

This crate requires Rust **1.32** or above.

## Usage

The interface provided by this crate is very basic. You supply a jq program
string and a string to run the program over.

```rust
use jq_rs;
// ...

let res = jq_rs::run(".name", r#"{"name": "test"}"#);
assert_eq!(res.unwrap(), "\"test\"\n".to_string());
```

In addition to running one-off programs with `jq_rs::run()`, you can also
use `jq_rs::compile()` to compile a jq program and reuse it with
different inputs.

```rust
use jq_rs;

let tv_shows = r#"[
 {"title": "Twilight Zone"},
 {"title": "X-Files"},
 {"title": "The Outer Limits"}
]"#;

let movies = r#"[
 {"title": "The Omen"},
 {"title": "Amityville Horror"},
 {"title": "The Thing"}
]"#;

let mut program = jq_rs::compile("[.[].title] | sort").unwrap();

assert_eq!(
 &program.run(tv_shows).unwrap(),
 "[\"The Outer Limits\",\"Twilight Zone\",\"X-Files\"]\n"
);

assert_eq!(
 &program.run(movies).unwrap(),
 "[\"Amityville Horror\",\"The Omen\",\"The Thing\"]\n",
);
```

## A Note on Performance

While the benchmarks are far from exhaustive, they indicate that much of the
runtime of a simple jq program goes to the compilation. In fact, the compilation
is _quite expensive_.

```text
run one off             time:   [48.594 ms 48.689 ms 48.800 ms]
Found 6 outliers among 100 measurements (6.00%)
3 (3.00%) high mild
3 (3.00%) high severe

run pre-compiled        time:   [4.0351 us 4.0708 us 4.1223 us]
Found 15 outliers among 100 measurements (15.00%)
6 (6.00%) high mild
9 (9.00%) high severe
```

If you have a need to run the same jq program multiple times it is
_highly recommended_ to retain a pre-compiled `JqProgram` and reuse it.

## Handling Output

The return values from jq are _strings_ since there is no certainty that the
output will be valid json. As such the output will need to be parsed if you want
to work with the actual data types being represented.

In such cases you may want to pair this crate with [serde_json] or similar.

For example, here we want to extract the numbers from a set of objects:

```rust
use jq_rs;
use serde_json::{self, json};

// ...

let data = json!({
 "movies": [
     { "title": "Coraline", "year": 2009 },
     { "title": "ParaNorman", "year": 2012 },
     { "title": "Boxtrolls", "year": 2014 },
     { "title": "Kubo and the Two Strings", "year": 2016 },
     { "title": "Missing Link", "year": 2019 }
 ]
});

let query = "[.movies[].year]";
// program output as a json string...
let output = jq_rs::run(query, &data.to_string()).unwrap();
// ... parse via serde
let parsed: Vec<i64> = serde_json::from_str(&output).unwrap();

assert_eq!(vec![2009, 2012, 2014, 2016, 2019], parsed);
```

Barely any of the options or flags available from the [jq] cli are exposed
currently.
Literally all that is provided is the ability to execute a _jq program_ on a blob
of json.
Please pardon my dust as I sort out the details.

## Linking to libjq

This crate requires access to `libjq` at build and/or runtime depending on the
your choice.

When the `bundled` feature is enabled (**off by default**) `libjq` is provided
and linked statically to your crate by [jq-sys] and [jq-src]. Using this feature
requires having autotools and gcc in `PATH` in order for the to build to work.

Without the `bundled` feature, _you_ will need to ensure your crate
can link to `libjq` in order for the bindings to work.

You can choose to compile `libjq` yourself, or perhaps install it via your
system's package manager.
See the [jq-sys building docs][jq-sys-building] for details on how to share
hints with the [jq-sys] crate on how to link.


[jq]: https://github.com/stedolan/jq
[serde_json]: https://github.com/serde-rs/json
[json-query]: https://crates.io/crates/json-query
[jq-sys]: https://github.com/onelson/jq-sys
[jq-sys-building]: https://github.com/onelson/jq-sys#building
[jq-src]: https://github.com/onelson/jq-src


CHANGELOG.md
# Changelog

## v0.4.1 ([2019-08-17](https://github.com/onelson/jq-rs/compare/v0.4.0..v0.4.1 "diff"))

Additions

- Implements `std::error::Error + Send + 'static` for `jq_rs::Error` to better
  integrate with popular error handling crate [error-chain] and others ([#22]).

## v0.4.0 ([2019-07-06](https://github.com/onelson/jq-rs/compare/v0.3.1..v0.4.0 "diff"))

Breaking Changes

- Renamed crate from `json-query` to `jq-rs` ([#12]).
- Adopted 2018 edition. The minimum supported rust version is now **1.32** ([#14]).
- Output from jq programs now includes a trailing newline, just like the output
  from the `jq` binary ([#6]).
- Added custom `Error` and `Result` types, returned from fallible
  functions/methods in this crate ([#8]).

## v0.3.1 ([2019-07-04](https://github.com/onelson/json-query/compare/v0.3.0..v0.3.1 "diff"))

**Note: This is final release with the name [json-query].
Future releases will be published as [jq-rs].**

Bugfixes

- Fixed issue where newlines in output were not being preserved correctly ([#3]).
- Resolved a memory error which could cause a crash when running a jq program
  which could attempt to access missing fields on an object ([#4]).
- Fixed some memory leaks which could occur during processing ([#10]).

## v0.3.0 ([2019-06-01](https://github.com/onelson/json-query/compare/v0.2.1..v0.3.0 "diff"))

- Added `json_query::compile()`. Compile a jq program, then reuse it, running
  it against several inputs.

## v0.2.1 ([2019-06-01](https://github.com/onelson/json-query/compare/v0.2.0..v0.2.1 "diff"))

- [#1] Enabled `bundled` feature when building on docs.rs.

## v0.2.0 ([2019-02-18](https://github.com/onelson/json-query/compare/v0.1.1..v0.2.0 "diff"))

- Updates [jq-sys] dep to v0.2.0 for better build controls.
- Settles on 2015 edition style imports (for now).

Breaking Changes:

- `bundled` feature is no longer enabled by default.


## v0.1.1 ([2019-01-14](https://github.com/onelson/json-query/compare/v0.1.0..v0.1.1 "diff"))

- Added extra links to cargo manifest.
- Added some basic docs.
- Added a `bundled` feature to opt in or out of using the bundled source.

## v0.1.0 (2019-01-13)

Initial release.

[jq]: https://github.com/stedolan/jq
[serde_json]: https://github.com/serde-rs/json
[jq-rs]: https://crates.io/crates/jq-rs
[json-query]: https://crates.io/crates/json-query
[jq-sys]: https://github.com/onelson/jq-sys
[jq-sys-building]: https://github.com/onelson/jq-sys#building
[jq-src]: https://github.com/onelson/jq-src
[error-chain]: https://crates.io/crates/error-chain

[#1]: https://github.com/onelson/json-query/issues/1
[#3]: https://github.com/onelson/json-query/issues/3
[#4]: https://github.com/onelson/json-query/issues/4
[#6]: https://github.com/onelson/jq-rs/pull/6
[#8]: https://github.com/onelson/jq-rs/pull/8
[#10]: https://github.com/onelson/json-query/issues/10
[#12]: https://github.com/onelson/jq-rs/issues/12
[#14]: https://github.com/onelson/jq-rs/issues/14
[#22]: https://github.com/onelson/jq-rs/pull/22


tests/error-chain-compat.rs
extern crate jq_rs;
#[macro_use]
extern crate error_chain;
use error_chain::ChainedError;

mod errors {
    error_chain! {
        foreign_links {
            Jq(jq_rs::Error);
        }
    }
}

use self::errors::{Error, ErrorKind, ResultExt};

#[test]
fn test_match_errorkind() {
    match jq_rs::run(".", "[[[{}}").unwrap_err().into() {
        Error(ErrorKind::Jq(e), _s) => {
            // Proving that jq_rs::Error does in fact implement `std::error::Error`.
            // Sort of redundant considering error_chain requires this, but it
            // can't hurt to say it explicitly here.
            use std::error::Error as _;
            let _ = e.source();
        }
        _ => unreachable!("error-chain should be converting."),
    }
}

#[test]
fn test_chain_err() {
    let chain = jq_rs::run(".", "[[[{}}")
        .chain_err(|| "custom message")
        .unwrap_err()
        .display_chain()
        .to_string();

    // the chain is a multi-line string mentioning each error in the chain.
    assert!(chain.contains("custom message"));
    assert!(chain.contains("Parse error"))
}


src/jq.rs
//! This module takes the unsafe bindings from `jq-sys` then (hopefully)
//! wrapping to present a slightly safer API to use.
//!
//! These are building blocks and not intended for use from the public API.

use crate::errors::{Error, Result};
use jq_sys::{
    jq_compile, jq_format_error, jq_get_exit_code, jq_halted, jq_init, jq_next, jq_set_error_cb,
    jq_start, jq_state, jq_teardown, jv, jv_copy, jv_dump_string, jv_free, jv_get_kind,
    jv_invalid_get_msg, jv_invalid_has_msg, jv_kind_JV_KIND_INVALID, jv_kind_JV_KIND_NUMBER,
    jv_kind_JV_KIND_STRING, jv_number_value, jv_parser, jv_parser_free, jv_parser_new,
    jv_parser_next, jv_parser_set_buf, jv_string_value,
};
use std::ffi::{CStr, CString};
use std::os::raw::{c_char, c_void};

pub struct Jq {
    state: *mut jq_state,
    err_buf: String,
}

impl Jq {
    pub fn compile_program(program: CString) -> Result<Self> {
        let mut jq = Jq {
            state: {
                // jq's master branch shows this can be a null pointer, in
                // which case the binary will exit with a `Error::System`.
                let ptr = unsafe { jq_init() };
                if ptr.is_null() {
                    return Err(Error::System {
                        reason: Some("Failed to init".into()),
                    });
                } else {
                    ptr
                }
            },
            err_buf: "".to_string(),
        };

        extern "C" fn err_cb(data: *mut c_void, msg: jv) {
            unsafe {
                let formatted = jq_format_error(msg);
                let jq = &mut *(data as *mut Jq);
                jq.err_buf += &(CStr::from_ptr(jv_string_value(formatted))
                    .to_str()
                    .unwrap_or("")
                    .to_string()
                    + "\n");
                jv_free(formatted);
            }
        }
        unsafe {
            jq_set_error_cb(jq.state, Some(err_cb), &mut jq as *mut Jq as *mut c_void);
        }

        if unsafe { jq_compile(jq.state, program.as_ptr()) } == 0 {
            Err(Error::InvalidProgram {
                reason: jq.err_buf.clone(),
            })
        } else {
            Ok(jq)
        }
    }

    fn is_halted(&self) -> bool {
        unsafe { jq_halted(self.state) != 0 }
    }

    fn get_exit_code(&self) -> ExitCode {
        let exit_code = JV {
            ptr: unsafe { jq_get_exit_code(self.state) },
        };

        // The rules for this seem odd, but I'm trying to model this after the
        // similar block in the jq `main.c`s `process()` function.

        if exit_code.is_valid() {
            ExitCode::JQ_OK
        } else {
            exit_code
                .as_number()
                .map(|i| (i as isize).into())
                .unwrap_or(ExitCode::JQ_ERROR_UNKNOWN)
        }
    }

    /// Run the jq program against an input.
    pub fn execute(&mut self, input: CString) -> Result<String> {
        let mut parser = Parser::new();
        self.process(parser.parse(input)?)
    }

    /// Unwind the parser and return the rendered result.
    ///
    /// When this results in `Err`, the String value should contain a message about
    /// what failed.
    fn process(&mut self, initial_value: JV) -> Result<String> {
        let mut buf = String::new();

        unsafe {
            // `jq_start` seems to be a consuming call.
            // In order to avoid a double-free, when `initial_value` is dropped,
            // we have to use `jv_copy` on the inner `jv`.
            jq_start(self.state, jv_copy(initial_value.ptr), 0);
            // After, we can manually free the `initial_value` with `drop` since
            // it is no longer needed.
            drop(initial_value);

            dump(self, &mut buf)?;
        }

        Ok(buf)
    }
}

impl Drop for Jq {
    fn drop(&mut self) {
        unsafe { jq_teardown(&mut self.state) }
    }
}

struct JV {
    ptr: jv,
}

impl JV {
    /// Convert the current `JV` into the "dump string" rendering of itself.
    pub fn as_dump_string(&self) -> Result<String> {
        let dump = JV {
            ptr: unsafe { jv_dump_string(jv_copy(self.ptr), 0) },
        };
        unsafe { get_string_value(jv_string_value(dump.ptr)) }
    }

    /// Attempts to extract feedback from jq if the JV is invalid.
    pub fn get_msg(&self) -> Option<String> {
        if self.invalid_has_msg() {
            let reason = {
                let msg = JV {
                    ptr: unsafe {
                        // This call is gross since we're dipping outside of the
                        // safe/drop-enabled wrapper to get a copy which will be freed
                        // by jq. If we wrap it in a `JV`, we'll run into a double-free
                        // situation.
                        jv_invalid_get_msg(jv_copy(self.ptr))
                    },
                };

                format!(
                    "JQ: Parse error: {}",
                    msg.as_string().unwrap_or_else(|_| "unknown".into())
                )
            };
            Some(reason)
        } else {
            None
        }
    }

    pub fn as_number(&self) -> Option<f64> {
        unsafe {
            if jv_get_kind(self.ptr) == jv_kind_JV_KIND_NUMBER {
                Some(jv_number_value(self.ptr))
            } else {
                None
            }
        }
    }

    pub fn as_string(&self) -> Result<String> {
        unsafe {
            if jv_get_kind(self.ptr) == jv_kind_JV_KIND_STRING {
                get_string_value(jv_string_value(self.ptr))
            } else {
                Err(Error::Unknown)
            }
        }
    }

    pub fn is_valid(&self) -> bool {
        unsafe { jv_get_kind(self.ptr) != jv_kind_JV_KIND_INVALID }
    }

    pub fn invalid_has_msg(&self) -> bool {
        unsafe { jv_invalid_has_msg(jv_copy(self.ptr)) == 1 }
    }
}

impl Drop for JV {
    fn drop(&mut self) {
        unsafe { jv_free(self.ptr) };
    }
}

struct Parser {
    ptr: *mut jv_parser,
}

impl Parser {
    pub fn new() -> Self {
        Self {
            ptr: unsafe { jv_parser_new(0) },
        }
    }

    pub fn parse(&mut self, input: CString) -> Result<JV> {
        // For a single run, we could set this to `1` (aka `true`) but this will
        // break the repeated `JqProgram` usage.
        // It may be worth exposing this to the caller so they can set it for each
        // use case, but for now we'll just "leave it open."
        let is_last = 0;

        // Originally I planned to have a separate "set_buf" method, but it looks like
        // the C api really wants you to set the buffer, then call `jv_parser_next()` in
        // the same logical block.
        // Mainly I think the important thing is to ensure the `input` outlives both the
        // set_buf and next calls.
        unsafe {
            jv_parser_set_buf(
                self.ptr,
                input.as_ptr(),
                input.as_bytes().len() as i32,
                is_last,
            )
        };

        let value = JV {
            ptr: unsafe { jv_parser_next(self.ptr) },
        };
        if value.is_valid() {
            Ok(value)
        } else {
            Err(Error::System {
                reason: Some(
                    value
                        .get_msg()
                        .unwrap_or_else(|| "JQ: Parser error".to_string()),
                ),
            })
        }
    }
}

impl Drop for Parser {
    fn drop(&mut self) {
        unsafe {
            jv_parser_free(self.ptr);
        }
    }
}

/// Takes a pointer to a nul term string, and attempts to convert it to a String.
unsafe fn get_string_value(value: *const c_char) -> Result<String> {
    let s = CStr::from_ptr(value).to_str()?;
    Ok(s.to_owned())
}

/// Renders the data from the parser and pushes it into the buffer.
unsafe fn dump(jq: &Jq, buf: &mut String) -> Result<()> {
    // Looks a lot like an iterator...

    let mut value = JV {
        ptr: jq_next(jq.state),
    };

    while value.is_valid() {
        let s = value.as_dump_string()?;
        buf.push_str(&s);
        buf.push('\n');

        value = JV {
            ptr: jq_next(jq.state),
        };
    }

    if jq.is_halted() {
        use ExitCode::*;
        match jq.get_exit_code() {
            JQ_ERROR_SYSTEM => Err(Error::System {
                reason: value.get_msg(),
            }),
            // As far as I know, we should not be able to see a compile error
            // this deep into the execution of a jq program (it would need to be
            // compiled already, right?)
            // Still, compile failure is represented by an exit code, so in
            // order to be exhaustive we have to check for it.
            JQ_ERROR_COMPILE => Err(Error::InvalidProgram {
                reason: jq.err_buf.clone(),
            }),
            // Any of these `OK_` variants are "success" cases.
            // I suppose the jq program can halt successfully, or not, or not at
            // all and still terminate some other way?
            JQ_OK | JQ_OK_NULL_KIND | JQ_OK_NO_OUTPUT => Ok(()),
            JQ_ERROR_UNKNOWN => Err(Error::Unknown),
        }
    } else if let Some(reason) = value.get_msg() {
        Err(Error::System {
            reason: Some(reason),
        })
    } else {
        Ok(())
    }
}

/// Various exit codes jq checks for during the `if (jq_halted(jq))` branch of
/// their processing loop.
///
/// Adapted from the enum seen in jq's master branch right now.
/// The numbers seem to line up with the magic numbers seen in
/// the 1.6 release, though there's no enum that I saw at that point in the git
/// history.
#[allow(non_camel_case_types, dead_code)]
#[rustfmt::skip]
enum ExitCode {
    JQ_OK               =  0,
    JQ_OK_NULL_KIND     = -1,
    JQ_ERROR_SYSTEM     =  2,
    JQ_ERROR_COMPILE    =  3,
    JQ_OK_NO_OUTPUT     = -4,
    JQ_ERROR_UNKNOWN    =  5,
}

impl From<isize> for ExitCode {
    #[rustfmt::skip]
    fn from(number: isize) -> Self {
        use ExitCode::*;
        match number {
             0 => JQ_OK,
            -1 => JQ_OK_NULL_KIND,
             2 => JQ_ERROR_SYSTEM,
             3 => JQ_ERROR_COMPILE,
            -4 => JQ_OK_NO_OUTPUT,
             // `5` is called out explicitly in the jq source, but also "unknown"
             // seems to make good sense for other unexpected number.
             5 | _ => JQ_ERROR_UNKNOWN,
        }
    }
}


src/errors.rs
use std::error;
use std::fmt;
use std::result;

const ERR_UNKNOWN: &str = "JQ: Unknown error";
const ERR_COMPILE: &str = "JQ: Program failed to compile";
const ERR_STRING_CONV: &str = "JQ: Failed to convert string";

/// This is the common Result type for the crate. Fallible operations will
/// return this.
pub type Result<T> = result::Result<T, Error>;

/// There are many potential causes for failure when running jq programs.
/// This enum attempts to unify them all under a single type.
#[derive(Debug)]
pub enum Error {
    /// The jq program failed to compile.
    InvalidProgram {
        /// JQ's explanation of the compilation error
        reason: String,
    },
    /// System errors are raised by the internal jq state machine. These can
    /// indicate problems parsing input, or even failures while initializing
    /// the state machine itself.
    System {
        /// Feedback from jq about what went wrong, when available.
        reason: Option<String>,
    },
    /// Errors encountered during conversion between CString/String or vice
    /// versa.
    StringConvert {
        /// The original error which lead to this.
        err: Box<dyn error::Error + 'static>,
    },
    /// Something bad happened, but it was unexpected.
    Unknown,
}

unsafe impl Send for Error {}

impl error::Error for Error {
    fn description(&self) -> &str {
        match self {
            Error::StringConvert { .. } => ERR_STRING_CONV,
            Error::InvalidProgram { reason } => reason,
            Error::System { reason } => reason
                .as_ref()
                .map(|x| x.as_str())
                .unwrap_or_else(|| ERR_UNKNOWN),
            Error::Unknown => ERR_UNKNOWN,
        }
    }

    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
        match self {
            Error::StringConvert { err } => {
                if let Some(err) = err.downcast_ref::<std::ffi::NulError>() {
                    Some(err)
                } else if let Some(err) = err.downcast_ref::<std::str::Utf8Error>() {
                    Some(err)
                } else {
                    None
                }
            }
            _ => None,
        }
    }
}

impl From<std::ffi::NulError> for Error {
    fn from(err: std::ffi::NulError) -> Self {
        Error::StringConvert { err: Box::new(err) }
    }
}

impl From<std::str::Utf8Error> for Error {
    fn from(err: std::str::Utf8Error) -> Self {
        Error::StringConvert { err: Box::new(err) }
    }
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let detail: String = match self {
            Error::InvalidProgram { reason } => format!("{}: {}", ERR_COMPILE, reason),
            Error::System { reason } => reason
                .as_ref()
                .cloned()
                .unwrap_or_else(|| ERR_UNKNOWN.into()),
            Error::StringConvert { err } => format!("{} - `{}`", ERR_STRING_CONV, err),
            Error::Unknown => ERR_UNKNOWN.into(),
        };
        write!(f, "{}", detail)
    }
}


benches/simple.rs
#[macro_use]
extern crate criterion;
extern crate jq_rs;

use criterion::black_box;
use criterion::Criterion;
use jq_rs::{JqProgram, Result};

fn run_one_off(prog: &str, input: &str) -> Result<String> {
    jq_rs::run(prog, input)
}

fn run_pre_compiled(prog: &mut JqProgram, input: &str) -> Result<String> {
    prog.run(input)
}

fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("run one off", |b| {
        b.iter(|| run_one_off(black_box(".name"), black_box(r#"{"name": "John Wick"}"#)))
    });

    c.bench_function("run pre-compiled", |b| {
        let mut prog = jq_rs::compile(".name").unwrap();
        b.iter(|| run_pre_compiled(black_box(&mut prog), black_box(r#"{"name": "John Wick"}"#)))
    });
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);


src/lib.rs
//! ## Overview
//!
//! > Prior to v0.4.0 this crate was named [json-query].
//!
//! This rust crate provides access to [jq] 1.6 via the `libjq` C API (rather than
//! "shelling out").
//!
//! By leveraging [jq] we can extract data from json strings using `jq`'s dsl.
//!
//! This crate requires Rust **1.32** or above.
//!
//! ## Usage
//!
//! The interface provided by this crate is very basic. You supply a jq program
//! string and a string to run the program over.
//!
//! ```rust
//! use jq_rs;
//! // ...
//!
//! let res = jq_rs::run(".name", r#"{"name": "test"}"#);
//! assert_eq!(res.unwrap(), "\"test\"\n".to_string());
//! ```
//!
//! In addition to running one-off programs with `jq_rs::run()`, you can also
//! use `jq_rs::compile()` to compile a jq program and reuse it with
//! different inputs.
//!
//! ```rust
//! use jq_rs;
//!
//! let tv_shows = r#"[
//!     {"title": "Twilight Zone"},
//!     {"title": "X-Files"},
//!     {"title": "The Outer Limits"}
//! ]"#;
//!
//! let movies = r#"[
//!     {"title": "The Omen"},
//!     {"title": "Amityville Horror"},
//!     {"title": "The Thing"}
//! ]"#;
//!
//! let mut program = jq_rs::compile("[.[].title] | sort").unwrap();
//!
//! assert_eq!(
//!     &program.run(tv_shows).unwrap(),
//!     "[\"The Outer Limits\",\"Twilight Zone\",\"X-Files\"]\n"
//! );
//!
//! assert_eq!(
//!     &program.run(movies).unwrap(),
//!     "[\"Amityville Horror\",\"The Omen\",\"The Thing\"]\n",
//! );
//! ```
//!
//! ## A Note on Performance
//!
//! While the benchmarks are far from exhaustive, they indicate that much of the
//! runtime of a simple jq program goes to the compilation. In fact, the compilation
//! is _quite expensive_.
//!
//! ```text
//! run one off             time:   [48.594 ms 48.689 ms 48.800 ms]
//! Found 6 outliers among 100 measurements (6.00%)
//!   3 (3.00%) high mild
//!   3 (3.00%) high severe
//!
//! run pre-compiled        time:   [4.0351 us 4.0708 us 4.1223 us]
//! Found 15 outliers among 100 measurements (15.00%)
//!   6 (6.00%) high mild
//!   9 (9.00%) high severe
//! ```
//!
//! If you have a need to run the same jq program multiple times it is
//! _highly recommended_ to retain a pre-compiled `JqProgram` and reuse it.
//!
//! ## Handling Output
//!
//! The return values from jq are _strings_ since there is no certainty that the
//! output will be valid json. As such the output will need to be parsed if you want
//! to work with the actual data types being represented.
//!
//! In such cases you may want to pair this crate with [serde_json] or similar.
//!
//! For example, here we want to extract the numbers from a set of objects:
//!
//! ```rust
//! use jq_rs;
//! use serde_json::{self, json};
//!
//! // ...
//!
//! let data = json!({
//!     "movies": [
//!         { "title": "Coraline", "year": 2009 },
//!         { "title": "ParaNorman", "year": 2012 },
//!         { "title": "Boxtrolls", "year": 2014 },
//!         { "title": "Kubo and the Two Strings", "year": 2016 },
//!         { "title": "Missing Link", "year": 2019 }
//!     ]
//! });
//!
//! let query = "[.movies[].year]";
//! // program output as a json string...
//! let output = jq_rs::run(query, &data.to_string()).unwrap();
//! // ... parse via serde
//! let parsed: Vec<i64> = serde_json::from_str(&output).unwrap();
//!
//! assert_eq!(vec![2009, 2012, 2014, 2016, 2019], parsed);
//! ```
//!
//! Barely any of the options or flags available from the [jq] cli are exposed
//! currently.
//! Literally all that is provided is the ability to execute a _jq program_ on a blob
//! of json.
//! Please pardon my dust as I sort out the details.
//!
//! ## Linking to libjq
//!
//! This crate requires access to `libjq` at build and/or runtime depending on the
//! your choice.
//!
//! When the `bundled` feature is enabled (**off by default**) `libjq` is provided
//! and linked statically to your crate by [jq-sys] and [jq-src]. Using this feature
//! requires having autotools and gcc in `PATH` in order for the to build to work.
//!
//! Without the `bundled` feature, _you_ will need to ensure your crate
//! can link to `libjq` in order for the bindings to work.
//!
//! You can choose to compile `libjq` yourself, or perhaps install it via your
//! system's package manager.
//! See the [jq-sys building docs][jq-sys-building] for details on how to share
//! hints with the [jq-sys] crate on how to link.
//!
//! [jq]: https://github.com/stedolan/jq
//! [serde_json]: https://github.com/serde-rs/json
//! [jq-rs]: https://crates.io/crates/jq-rs
//! [json-query]: https://crates.io/crates/json-query
//! [jq-sys]: https://github.com/onelson/jq-sys
//! [jq-sys-building]: https://github.com/onelson/jq-sys#building
//! [jq-src]: https://github.com/onelson/jq-src

#![deny(missing_docs)]

extern crate jq_sys;
#[cfg(test)]
#[macro_use]
extern crate serde_json;

mod errors;
mod jq;

use std::ffi::CString;

pub use errors::{Error, Result};

/// Run a jq program on a blob of json data.
///
/// In the case of failure to run the program, feedback from the jq api will be
/// available in the supplied `String` value.
/// Failures can occur for a variety of reasons, but mostly you'll see them as
/// a result of bad jq program syntax, or invalid json data.
pub fn run(program: &str, data: &str) -> Result<String> {
    compile(program)?.run(data)
}

/// A pre-compiled jq program which can be run against different inputs.
pub struct JqProgram {
    jq: jq::Jq,
}

impl JqProgram {
    /// Runs a json string input against a pre-compiled jq program.
    pub fn run(&mut self, data: &str) -> Result<String> {
        if data.trim().is_empty() {
            // During work on #4, #7, the parser test which allows us to avoid a memory
            // error shows that an empty input just yields an empty response BUT our
            // implementation would yield a parse error.
            return Ok("".into());
        }
        let input = CString::new(data)?;
        self.jq.execute(input)
    }
}

/// Compile a jq program then reuse it, running several inputs against it.
pub fn compile(program: &str) -> Result<JqProgram> {
    let prog = CString::new(program)?;
    Ok(JqProgram {
        jq: jq::Jq::compile_program(prog)?,
    })
}

#[cfg(test)]
mod test {

    use super::{compile, run, Error};
    use matches::assert_matches;
    use serde_json;

    #[test]
    fn reuse_compiled_program() {
        let query = r#"if . == 0 then "zero" elif . == 1 then "one" else "many" end"#;
        let mut prog = compile(&query).unwrap();
        assert_eq!(prog.run("2").unwrap(), "\"many\"\n");
        assert_eq!(prog.run("1").unwrap(), "\"one\"\n");
        assert_eq!(prog.run("0").unwrap(), "\"zero\"\n");
    }

    #[test]
    fn jq_state_is_not_global() {
        let input = r#"{"id": 123, "name": "foo"}"#;
        let query1 = r#".name"#;
        let query2 = r#".id"#;

        // Basically this test is just to check that the state pointers returned by
        // `jq::init()` are completely independent and don't share any global state.
        let mut prog1 = compile(&query1).unwrap();
        let mut prog2 = compile(&query2).unwrap();

        assert_eq!(prog1.run(input).unwrap(), "\"foo\"\n");
        assert_eq!(prog2.run(input).unwrap(), "123\n");
        assert_eq!(prog1.run(input).unwrap(), "\"foo\"\n");
        assert_eq!(prog2.run(input).unwrap(), "123\n");
    }

    fn get_movies() -> serde_json::Value {
        json!({
            "movies": [
                { "title": "Coraline", "year": 2009 },
                { "title": "ParaNorman", "year": 2012 },
                { "title": "Boxtrolls", "year": 2014 },
                { "title": "Kubo and the Two Strings", "year": 2016 },
                { "title": "Missing Link", "year": 2019 }
            ]
        })
    }

    #[test]
    fn identity_nothing() {
        assert_eq!(run(".", "").unwrap(), "".to_string());
    }

    #[test]
    fn identity_empty() {
        assert_eq!(run(".", "{}").unwrap(), "{}\n".to_string());
    }

    #[test]
    fn extract_dates() {
        let data = get_movies();
        let query = "[.movies[].year]";
        let output = run(query, &data.to_string()).unwrap();
        let parsed: Vec<i64> = serde_json::from_str(&output).unwrap();
        assert_eq!(vec![2009, 2012, 2014, 2016, 2019], parsed);
    }

    #[test]
    fn extract_name() {
        let res = run(".name", r#"{"name": "test"}"#);
        assert_eq!(res.unwrap(), "\"test\"\n".to_string());
    }

    #[test]
    fn unpack_array() {
        let res = run(".[]", "[1,2,3]");
        assert_eq!(res.unwrap(), "1\n2\n3\n".to_string());
    }

    #[test]
    fn compile_error() {
        let res = run(". aa12312me  dsaafsdfsd", "{\"name\": \"test\"}");
        assert_matches!(res, Err(Error::InvalidProgram { .. }));
    }

    #[test]
    fn parse_error() {
        let res = run(".", "{1233 invalid json ahoy : est\"}");
        assert_matches!(res, Err(Error::System { .. }));
    }

    #[test]
    fn just_open_brace() {
        let res = run(".", "{");
        assert_matches!(res, Err(Error::System { .. }));
    }

    #[test]
    fn just_close_brace() {
        let res = run(".", "}");
        assert_matches!(res, Err(Error::System { .. }));
    }

    #[test]
    fn total_garbage() {
        let data = r#"
        {
            moreLike: "an object literal but also bad"
            loveToDangleComma: true,
        }"#;

        let res = run(".", data);
        assert_matches!(res, Err(Error::System { .. }));
    }

    pub mod mem_errors {
        //! Attempting run a program resulting in bad field access has been
        //! shown to sometimes trigger a use after free or double free memory
        //! error.
        //!
        //! Technically the program and inputs are both valid, but the
        //! evaluation of the program causes bad memory access to happen.
        //!
        //! https://github.com/onelson/json-query/issues/4

        use super::*;

        #[test]
        fn missing_field_access() {
            let prog = ".[] | .hello";
            let data = "[1,2,3]";
            let res = run(prog, data);
            assert_matches!(res, Err(Error::System { .. }));
        }

        #[test]
        fn missing_field_access_compiled() {
            let mut prog = compile(".[] | .hello").unwrap();
            let data = "[1,2,3]";
            let res = prog.run(data);
            assert_matches!(res, Err(Error::System { .. }));
        }
    }
}
